### ISSUE TITLE
Investigating ESLint migration

### ISSUE THREAD

**Issue by 43081j** [microsoft/dtslint#300]
Hello 👋 

Recently, out of interest, I made a branch of dtslint which drops tslint entirely and uses ESLint instead.

TSLint has been obsolete for a very long time, and deprecated for quite some time too. It would be good to move towards removing it IMO, is that something you're interested in?

I'm happy to continue my branch if that's the case, but it would be very nice to have some help converting the individual rules as there are many and I haven't touched this repo before.

Couple of notes from what I've found so far:

* dts-critic returns a diagnostic which has a position of `{start, length}` which is aligned with TSLint but not ESLint. In ESLint we need a `{column, line}` or a `Node`
* Rules can be converted easily if you understand their logic already, this is where I am slow since I have to first understand each rule's intention and what it does
* We used to interact with the typescript api directly, but it seems this might go away as you can simply rely on ESLint to have already parsed/processed the AST instead
* For those rules that need type information, if any, we can use typescript-eslint's additional types/helpers

If you're interested in seeing my branch, i could make a draft PR which doesn't yet build and leave comments where I'm unsure what to do (usually vague types that were previously `any` or inferred).

**Comment by orta** [comment]
Hi! Sorry it took a while for us to respond. 

We've also been thinking about this off and on, and are definitely open to getting it working and replacing tslint! I've pushed a new branch which we can work towards: https://github.com/microsoft/dtslint/tree/eslint_base

**Comment by 43081j** [comment]
do you want me to open a draft PR against that? its been a while so ill catch it up from master too while im in there.

then we can decide if the stuff i already did is worth using as a starting point or not

**Comment by orta** [comment]
Yeah, that'd be great!

**Comment by JoshuaKGoldberg** [comment]
:wave: is anything happening here? I'd be interested in pitching in!

**Comment by 43081j** [comment]
@JoshuaKGoldberg it is being tracked in #325 by @sandersn , although he's presumably been preoccupied with other stuff recently. But reviewing that pr would be a good start I think

**Comment by 43081j** [comment]
#325 is stale as sanders doesn't currently have the time to push ahead with it, so any takers for opening a new PR would be great.

some feedback from me:

* we should probably see if there's a sensible way to migrate a rule at a time, one PR per rule
* similarly investigate how we can run both linters alongside each other when necessary
* IMO we should start with little or no eslint plugins to avoid pulling in bloat, ideally this repo can ultimately behave like an eslint plugin in the end

**Comment by JoshuaKGoldberg** [comment]
Yeah I'm planning on looking at this within the next week or so. My rough thoughts are that it would be cleanest long term to promote `eslint-plugin-expect-type`, so that folks don't need a separate linting tool other than ESLint.

From the perspective of DefinitelyTyped it would likely be very difficult to immediately switch from ESLint to TSLint given the sheer scale of the codebase and how many equivalent rules have subtle differences between the two linters. My first guess for a strategy would be something like:
1. Add ESLint + typescript-eslint for explicitly just the expect-type ruleset, removing the need for most or all id dtslint
  * Use typescript-eslint's TSLint runner to run TSLint inside ESLint
2. Switch over rules one by one from TSLint to ESLint
3. Remove TSLint altogether

...but that's just a rough guess that should be validated 😄

**Comment by 43081j** [comment]
keep in mind dtslint has plenty of useful rules beyond type assertions. that is actually only one rule, a small amount of the source of dtslint.

if there's enough complexity in that one rule to justify pulling in third party libs, fair enough.

the rest can mostly be converted fairly easily (but time consuming). so it would be super nice if we could have them running alongside each other which i think is what you're saying. so we can migrate a rule at a time, leaving the rest to continue using tslint.

i also already have a branch with some of those rules converted but none of the surrounding architecture updated to call them properly. so let me know if you want any of that in your efforts

**Comment by JoshuaKGoldberg** [comment]
I've reached out to @ibezkrovnyi to talk about `eslint-plugin-expect-type`. Looking through https://github.com/microsoft/dtslint/tree/76886d9/docs, indeed, there are quite a few rules that don't make sense under an "expect type" umbrella. I wonder if changing to a name like `eslint-plugin-types` is a reasonable future?

**Comment by sandersn** [comment]
I don't know the eslint naming scheme, but the rules are really style rules for Definitely Typed. I'd say that (at least) a few of them don't make sense outside that context, like the lint rules about when to use `declare` vs `export` in a d.ts file. Does eslint-plugin-definitely-typed make sense? That's just a guess.

**Comment by JoshuaKGoldberg** [comment]
Maybe two packages then, `eslint-plugin-types` and `eslint-plugin-definitelytyped`?

My main goal here is extracting the parts of dtslint that are broadly applicable into a standard ESLint plugin. My secondary goal is making dtslint use a typical ESLint plugin architecture so it can be made portable.

**Comment by 43081j** [comment]
Yeah I mean there's a good amount of purely dt-specific rules that should continue to live in this repo (even if this repo becomes an eslint plugin). But it's true a couple of them could be useful to other people, through even then someone could just use this plugin (eslint-plugin-dts for example). 

I've also converted a fair few of the rules already to eslint rules. if someone gets time to convert the architecture in general I could easily pr them in

**Comment by JoshuaKGoldberg** [comment]
FWIW I've taken on ownership of [`eslint-plugin-expect-type`](https://github.com/JoshuaKGoldberg) and would be happy to fold everything into a consolidated repo. I'll try to find time to look at converting architecture in general Soon™️!

**Comment by 43081j** [comment]
so would the plan be to update dtslint to make use of the expect-type plugin? given that it has many more DT-specific rules that don't belong in such a plugin (but rather in a DT-specific plugin, like this repo right here)

**Comment by JoshuaKGoldberg** [comment]
Tentatively, my proposal is that we convert **this** repo into a monorepo containing at least three packages:

* `eslint-plugin-types`: essentially what is `eslint-plugin-expect-type` today, with the addition of more general-use type rules not specifically related to type expectations 
* `eslint-plugin-definitelytyped`: rules specific to DefinitelyTyped and equivalent type definitions
    * Maybe this should be `eslint-plugin-dts`? If any rules exist _exclusively_ for just the DT repo, maybe they should live in that repo?
* `dtslint`: wrapper around ESLint that behaves equivalently to `dtslint` today

**Comment by JoshuaKGoldberg** [comment]
Tentatively, my proposal is that we convert **this** repo into a monorepo containing at least three packages:

* `eslint-plugin-types`: essentially what is `eslint-plugin-expect-type` today, with the addition of more general-use type rules not specifically related to type expectations 
* `eslint-plugin-definitelytyped`: rules specific to DefinitelyTyped and equivalent type definitions
    * Maybe this should be `eslint-plugin-dts`? If any rules exist _exclusively_ for just the DT repo, maybe they should live in that repo?
* `dtslint`: wrapper around ESLint that behaves equivalently to `dtslint` today

In theory we could use the separate `eslint-plugin-expect-type` repo as a source for the type expect rules. A few drawbacks IMO:
* We'd now have >=2 repositories' worth of tooling to maintain, despite this one likely wanting to be a monorepo anyway (to split the `dtslint` CLI and its community-applicable rules)
* This repository has a lot more eyes on it & previous bugs + features filed
* It'd be nice to bring in type assertions a little closer to TypeScript organizationally, since we can't build them into the language 😉

**Comment by 43081j** [comment]
that makes sense i think, especially maintaining this repo and just shuffling it around.

i already converted a fair few of the rules to eslint rules roughly (untested) so would be happy to contribute those if we did make some kind of `eslint-plugin-dts`

i think however we reorganise it, though, the big job is migrating dtslint to run eslint rather than tslint. the rules are already essentially lint rules, so pulling those out into a separate package isn't a blocker to migrating the core imo. may be worth starting to look at that before anything else

**Comment by JoshuaKGoldberg** [comment]
Poking through #325's changes and some of the adjacent areas of code, there are a few logic paths in `dtslint` that are specific to DefinitelyTyped.

* **`checkTslintJson`**: asserting that a `tslint.json` exists if a package is inferred as a DT one
    * Proposal: move that into a standalone piece of logic in DefinitelyTyped itself?
* **`assertPathIsInDefinitelyTyped`**: if a file contains a `// Type Definition` header, it should be inside DefinitelyTyped
    * Proposal: switch that into a general-use rule as a part of a recommended `eslint-plugin-dts` preset?
* **`updateConfig`**: standalone script to update TSLint configs for DefinitelyTyped
    * Proposal: move that into a standalone piece of logic in DefinitelyTyped itself?
    * It's basically https://github.com/amanda-mitchell/suppress-eslint-errors, but for dtslint

cc @sandersn

**Comment by sandersn** [comment]
* checkTslintJson/testDependencies: Maybe better moved to a new `@definitelytyped/dtslint-utils` package, and then called from dtslint? `@definitelytyped/dtslint-runner` is the package that calls dtslint, but it's not quite the right place for testDependencies since it's just supposed to determine which packages to call dtslint on. (There's surely a better name than `dtslint-utils`, of course.)
* updateConfig: Yes, move to DT under `scripts/`
* assertPathIsInDefinitelyTyped: Yes, should be a lint rule.

I like the monorepo idea, although it's a bit of work to set up. Unless it would be tremendously difficult, it'd be better to keep it here and replace the existing structure.

My main piece of advice, having recently upgraded all of DT at once to support exactOptionalPropertyTypes, is to make the transition incremental. Have dtslint run both tslint and eslint, and switch one rule over at a time. That'll make it easier to alternate between switching a lint rule and making sure all the DT packages still pass.

**Comment by sandersn** [comment]
Update: @andrewbranch and I talked about problems we've had updating dtslint with the rest of the `@definitelytyped` tools. We think it would make sense to move the bulk of dtslint into the `@definitelytyped` monorepo and either (1) publish a small wrapper for non-DT use from this repo or (2) keep publishing the package as dtslint from inside the monorepo.

Our motivation is the dependency `@definitelytyped/utils -> dtslint -> @definitelytyped/dtslint-runner`, which means that the `@definitelytyped` monorepo ends up with dependencies on itself in its yarn.lock.

Does that make sense? @JoshuaKGoldberg it's basically what you're doing, but even more so.

**Comment by 43081j** [comment]
dtslint contains rules only useful to DT, though. There's only maybe one or two rules appropriate for a "community" plugin.

It still is a tool specific to DT and exists to help maintain that repo and nothing else. I feel like because of that, dtslint should continue to exist whether it's in DT repo or not. 

Even if you split out the expect type rule, that's a really small part of a bigger tool.

**Comment by JoshuaKGoldberg** [comment]
👍  I have been convinced, thanks for the detailed explanations! Keeping DT specific runtime to the DT repo makes sense.

**Comment by sandersn** [comment]
> do you still want assistance in moving to ESLint internally

Yes, although it makes sense to move dtslint into DefinitelyTyped-tools first (in my opinion), and a completed eslint-plugin-expect-type might replace dtslint's expectRule.

If nobody else gets to it first, I'll probably work on the dtslint move during my next DT rotation, some time in November I think.

**Comment by loynoir** [comment]
Hi, guys. Newbie question here.

I saw some PR with title `eslint`.
Is `dtslint` somehow usable within `eslint` now ?

PS: I used `tsd`, works great, but seems no editor plugin yet.

**Comment by loynoir** [comment]
typescript < 4.5, else not working

https://www.npmjs.com/package/eslint-plugin-dtslint

```sh
$ tail dtstest.ts dtslint.eslintrc.cjs tsconfig.json
==> dtstest.ts <==
const answer = 42; // $ExpectType 41

==> dtslint.eslintrc.cjs <==
const { join } = require("path");
module.exports = {
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 2019,
    project: join(__dirname, "./tsconfig.json"),
    sourceType: "module"
  },
  extends: ["plugin:dtslint/recommended"],
};

==> tsconfig.json <==
{
  "extends": "@tsconfig/node14"
}
$ eslint -c dtslint.eslintrc.cjs dtstest.ts 

  1:1  error  Expected type to be: 41; got: 42  dtslint/expect-type

✖ 1 problem (1 error, 0 warnings)
```
